<a href="https://colab.research.google.com/github/Gracesugut/SQL-practice/blob/master/SQL_for_data_science_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#importing pandas
import pandas as pd
#loading sql extension
%load_ext sql
#connecting to sql lite database
%sql sqlite://

'Connected: @None'

In [ ]:
#import and preview of the dataset
accidents=pd.read_csv('traffic_accidents.csv')
accidents.head()

crash_unit_id     rd_no  ... mcs_out_of_service_i  hazmat_class
0         758369  JC469557  ...                  NaN           NaN
1         758356  JC469523  ...                  NaN           NaN
2         758329  JC469467  ...                  NaN           NaN
3         758330  JC469467  ...                  NaN           NaN
4         758333  JC469482  ...                  NaN           NaN

[5 rows x 71 columns]

In [ ]:
#delete the table if it exists in our database
%sql DROP TABLE if EXISTS accidents;
#Create table in our database
%sql PERSIST accidents;

 * sqlite://
Done.
 * sqlite://


'Persisted accidents'

In [ ]:
# view first five rows in the table 
%sql SELECT * FROM accidents LIMIT 5;

 * sqlite://
Done.


[(0, 758369, 'JC469557', '2019-10-12T01:57:00', 1, 'DRIVER', None, 720366, None, 'DODGE', 'NITRO', 'IL', 2011, 'UNKNOWN', 'PASSENGER', 'UNKNOWN/NA', 'W', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 1, None, 'OTHER', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (1, 758356, 'JC469523', '2019-10-12T00:30:00', 1, 'DRIVER', None, 720356, None, 'VOLVO', 'XC90', 'IL', 2020, 'NONE', 'SPORT UTILITY VEHICLE (SUV)', 'PERSONAL', 'S', 'SLOW/STOP IN TRAFFIC', None, None, 1, None, None, None, None, None, None, 1, None, None, None, None, None, None, None, None, None, None, 'SIDE-RIGHT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (2, 758329, 'JC469467', '2019-10-11T23:41:00', 1, 'DRIVER', 2.0, 720325, None, 'TOYOTA MOTOR COMPANY, LTD.', 'UNKNOWN', 'IL', 2015, 'NONE', 'PASSENGER', 'PERSONAL', 'E', 'PASSING/OVERTAKING', 1, None, 3, None, 'unknown', 'unknown', None, None, None, 1, 1, 1, 1, 1, None, None, None, None, None, None, 'REAR', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (3, 758330, 'JC469467', '2019-10-11T23:41:00', 2, 'DRIVER', None, 720334, None, 'SUBARU', 'FORESTER', 'IL', 2014, 'NONE', 'PASSENGER', 'PERSONAL', 'E', 'TURNING LEFT', None, None, 1, None, None, None, None, 1, 1, None, None, None, None, None, None, None, None, None, 1, None, 'FRONT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (4, 758333, 'JC469482', '2019-10-11T23:20:00', 1, 'DRIVER', None, 720328, None, 'CHEVROLET', 'IMPALA', 'IL', 2017, 'UNKNOWN', 'PASSENGER', 'PERSONAL', 'N', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'UNKNOWN', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None)]

**Display a list of 5 makes that had the highest number of crashes**

In [ ]:
#Display a list of 5 makes that had the highest number of crashes
%%sql 
SELECT make,COUNT(make) as accident_frequency   FROM accidents 
GROUP BY make 
ORDER BY accident_frequency DESC
LIMIT 5;


 * sqlite://
Done.


[('TOYOTA MOTOR COMPANY, LTD.', 637),
 ('CHEVROLET', 573),
 ('FORD', 566),
 ('NISSAN', 491),
 ('HONDA', 369)]

*Toyota Motor Company* had the highest number of accidents with *637* accidents followed *Chevrolet,Ford,Nissan and Honda* with *573, 566, 491 and  369* accidents respectively

**Most personal vehicles involved in crashes were travelling towards which direction?**

In [ ]:
#Most personal vehicles involved in crashes were travelling towards which direction? 
#
%%sql
SELECT  travel_direction ,COUNT(vehicle_use) as personal_vehicles_frequency FROM accidents
WHERE vehicle_use='PERSONAL'
GROUP BY travel_direction
ORDER BY personal_vehicles_frequency DESC;

 * sqlite://
Done.


[('N', 885),
 ('S', 852),
 ('W', 807),
 ('E', 743),
 ('UNKNOWN', 118),
 ('SE', 52),
 ('NW', 49),
 ('SW', 32),
 ('NE', 26)]

personal vehicles heading *North* were involved in most accidents with a total of *885* accidents 

**How many passengers were involved in personal cars were in accidents?**

In [ ]:
#How many passengers were involved in personal cars were in accidents?
%%sql
SELECT  SUM(num_passengers) as total_passangers FROM accidents
WHERE vehicle_use='PERSONAL';

 * sqlite://
Done.


[(869.0,)]

A total of *869*  passangers were in accidents involving personal cars

**How many passengers were involved in accidents while travelling in personal Sport Utility vehicles?**

In [ ]:
# How many passengers were involved in accidents while travelling in personal Sport Utility vehicles?
%%sql
SELECT  SUM(num_passengers) as total_passangers FROM accidents
WHERE vehicle_type='SPORT UTILITY VEHICLE (SUV)';

 * sqlite://
Done.


[(174.0,)]

A total of *174*  passangers were in accidents involving *sport utility vehicles*

**Compute the total number of Mercedez-Benz vehicles involved in accidents while being parked.**

In [ ]:
#Compute the total number of Mercedez-Benz vehicles involved in accidents while being parked.
%%sql
SELECT COUNT(make) as no_of_mercedez_accidents_parked FROM accidents
WHERE make='MERCEDES-BENZ' and maneuver='PARKED'

 * sqlite://
Done.


[(10,)]

*10 mercedez benz vehicles* were involved in accidents while *parked*

**Compute the total occupant count per vehicle count.**

In [ ]:
#Compute the total occupant count per vehicle count.
%%sql
SELECT make,AVG(occupant_cnt) as Total_occupant_count_per_vehicle_count 
FROM accidents
GROUP BY make
ORDER BY Total_occupant_count_per_vehicle_count DESC;

 * sqlite://
Done.


[("FLEET CAP'N TRAILERS", 4.0),
 ('PREVOST CAR, BUS, SAINTE CLAIRE, QUEBEC', 4.0),
 ('ICL CARGO TRAILER', 3.75),
 ('FIESTA (IMPORTED BY FORD)', 3.0),
 ('AMTRAN CORP, GENSIS BUSES', 2.3333333333333335),
 ('BLUE BIRD BODY CO.', 2.3333333333333335),
 ('UNKNOWN', 2.088888888888889),
 ('WABCO CONSTRUCTION & MINING EQUIPMENT GROUP', 2.0),
 ('FIAT', 1.6666666666666667),
 ('AUTOCAR', 1.5),
 ('MINI', 1.4444444444444444),
 ('NEW FLYER', 1.4390243902439024),
 ('BLUEBIRD INTERNATIONAL', 1.4),
 ('KAWASAKI MOTORS CORP.', 1.3333333333333333),
 ('CHRYSLER', 1.3097345132743363),
 ('DODGE', 1.251063829787234),
 ('HINO', 1.25),
 ('INTERNATIONAL HARVESTER', 1.25),
 ('BUICK', 1.205128205128205),
 ('LINCOLN-CONTINENTAL', 1.2),
 ('PONTIAC', 1.2),
 ('JAGUAR', 1.1666666666666667),
 ('KIA MOTORS CORP', 1.1644736842105263),
 ('NISSAN', 1.1629327902240325),
 ('SUBARU', 1.1475409836065573),
 ('FORD', 1.1360424028268552),
 ('MITSUBISHI', 1.1333333333333333),
 ('JEEP', 1.123931623931624),
 ('SUZUKI', 1.1111111111111112),
 ('LEXUS', 1.1014492753623188),
 ('ISUZU', 1.1),
 ('SATURN', 1.1),
 ('GENERAL MOTORS CORP.', 1.0985915492957747),
 ('TOYOTA MOTOR COMPANY, LTD.', 1.0910518053375196),
 ('AUDI', 1.0909090909090908),
 ('CHEVROLET', 1.087260034904014),
 ('HYUNDAI', 1.0833333333333333),
 ('HONDA', 1.0785907859078592),
 ('BMW', 1.0740740740740742),
 ('SCION', 1.0588235294117647),
 ('MAZDA', 1.0481927710843373),
 ('VOLVO', 1.0333333333333334),
 ('(UNLISTED MANUFACTURER)', 1.0),
 ('AMERICAN TRAILER & MFG. CO. (SEE ALSO TRANSPORTER)', 1.0),
 ('ASTON-MARTIN', 1.0),
 ('BANTAM CRANE', 1.0),
 ('CADILLAC', 1.0),
 ('CALDWELL DIV., SMITH INTERNATIONAL, INC.', 1.0),
 ('CASCADE COACH CO.', 1.0),
 ('CHEVELLE MOBILE HOMES, INC.', 1.0),
 ('CONTINENTAL', 1.0),
 ('EIDAL INTERNATIONAL CORP.', 1.0),
 ('EL DORADO INTERNATIONAL, INC.; EL DORADO MFG., LTD.', 1.0),
 ('FLEETCRAFT CORP.', 1.0),
 ('FLEETWOOD ENTERPRISES', 1.0),
 ('FONTAINE TRUCK EQUIPMENT CO.', 1.0),
 ('GILL MFG. CO.', 1.0),
 ('GOODYEAR CAMPER MFG., INC., LINCOLN, NB', 1.0),
 ('HAULER-TRAILER(MFD. BY NORDINE MFG.)', 1.0),
 ('HINDUSTAN', 1.0),
 ('INFINITI', 1.0),
 ('INTENATIONAL INDUSTRIES, INC.', 1.0),
 ('INTERNATIONAL COACH MFG.', 1.0),
 ('INTERNATIONAL TRAILER CORP.', 1.0),
 ('LINCOLN ELECTRIC CO., INC. CLEVELAND, OH ; LINCOLN WELDER', 1.0),
 ('MACK TRAILER MFG. CO., INC., DETRIOT, MI', 1.0),
 ('MACK TRUCKS, INC.', 1.0),
 ('MOTOR COACH INDUSTRIES, INC.', 1.0),
 ('NAVISTAR', 1.0),
 ('NISSAN DIESEL MOTOR CO.', 1.0),
 ('OLDSMOBILE', 1.0),
 ('PORSCHE', 1.0),
 ('ROLLS INTERNATIONAL TRAILER(MOBILE HOME)', 1.0),
 ('SAAB', 1.0),
 ('SCOOTALONG', 1.0),
 ('STERLINE FLATBED TRAILER', 1.0),
 ('STERLING', 1.0),
 ('TESLA MOTORS', 1.0),
 ('THOMAS BUILT BUS CO.', 1.0),
 ('TOYOCAR VAN CONTAINER TRAILER', 1.0),
 ('WORK HORSE MFG.', 1.0),
 ('MERCEDES-BENZ', 0.9852941176470589),
 ('VOLKSWAGEN', 0.9823008849557522),
 ('FREIGHTLINER CORP.', 0.9464285714285714),
 ('ACURA (DIV. OF AMERICAN HONDA MOTOR CO.)', 0.9375),
 ('ACURA', 0.9166666666666666),
 ('MERCURY', 0.9),
 ('KENWORTH MOTOR TRUCK CO.', 0.8333333333333334),
 ('AM GENERAL/HUMMER', 0.75),
 ('PETERBILT MOTORS CO., DIVISION PACCAR, INC.', 0.75),
 ('RANGE ROVER OF NORTH AMERICA', 0.75),
 ('LAND ROVER', 0.7058823529411765),
 ('HARLEY-DAVIDSON', 0.6),
 ('GREAT DANE TRAILERS INC.', 0.5),
 ('YAMAHA', 0.5),
 ('ALFA ROMEO', 0.0),
 ('BENTLEY', 0.0),
 ('FERRARI', 0.0),
 ('QUALITY TRAILER', 0.0),
 ('TOWMASTER', 0.0),
 ('TRIUMPH', 0.0)]

 **Sort the number of passengers in descending order by travel direction taking into account travel direction. Return travel direction and no. of passengers.**

In [ ]:
%%sql
SELECT travel_direction, SUM(num_passengers) as no_of_passangers from accidents
GROUP BY travel_direction
ORDER BY no_of_passangers DESC;

 * sqlite://
Done.


[('W', 297.0),
 ('N', 295.0),
 ('E', 283.0),
 ('S', 283.0),
 ('SE', 23.0),
 ('SW', 18.0),
 ('UNKNOWN', 14.0),
 ('NE', 8.0),
 ('NW', 8.0)]

Majority of the passangers involved in accidents were headed *west* with a total of *297* passangers followed by *North, East, South, South East, South West ,Unkown destination, North East and North West* with *295, 283, 283, 23, 18, 14, 8, 8* respectively

**Select top 5 vehicle models involved in the accident with no defect, order them by make and show their manoeuvre.** 

In [ ]:
%%sql
SELECT make,model,maneuver,COUNT(make) as no_of_vehicles FROM accidents
WHERE vehicle_defect= 'NONE'
GROUP BY model
ORDER BY no_of_vehicles DESC
LIMIT 5;

 * sqlite://
Done.


[('MERCEDES-BENZ', 'UNKNOWN', 'STRAIGHT AHEAD', 405),
 ('TOYOTA MOTOR COMPANY, LTD.', 'CAMRY', 'TURNING RIGHT', 135),
 ('NISSAN', 'NISSAN ALTIMA', 'SLOW/STOP IN TRAFFIC', 82),
 ('NISSAN', 'SENTRA (DATSUN AND NISSAN HAVE MERGED)', 'STRAIGHT AHEAD', 72),
 ('CHEVROLET', 'MALIBU (CHEVELLE)', 'PARKED', 69)]

Most of the vehicles involved in accident were of *mercedez benz* make unknown model with *405 vehicles* followed by *Toyota Camry, Nissan Altima,Nissan Sentra and Chevrolet*  with *135, 82, 72, 69* respectively 

**Which ridesharing vehicles in an accident had the most number of passengers? Return model, and no of passengers**

In [ ]:
%%sql
SELECT model,SUM(num_passengers) as No_of_passangers FROM accidents
WHERE vehicle_use='RIDESHARE SERVICE'
GROUP BY model
ORDER BY No_of_passangers DESC
LIMIT 5;

 * sqlite://
Done.


[('COROLLA', 5.0),
 ('MKZ', 5.0),
 ('CAMRY', 4.0),
 ('Prius', 4.0),
 ('UNKNOWN', 4.0)]

*Corolla and MKZ model* from the rideshare service had the most number of passangers with 5 passangers each followed by *Camry, Prius and unknown models* with *4* passangers each

## **Recomendations**

**Recommendation 1:**Toyota Motor Company had the highest number of accidents with 637 accidents followed Chevrolet,Ford,Nissan and Honda with 573, 566, 491 and 369 accidents respectively. Therefore resolve to having checks on car makes with the highest accident frequency to ensure road worthyness.

**Reccomendation 2:** Personal vehicles heading North, South,west, East had high accident frequency of 885, 852, 807, 743 respectively thus resolve to checking the state of roads along the specified areas 

**Reccomendation 3:**Most of the vehicles with no defects involved in accident were of mercedez benz make unknown model with 405 vehicles followed by Toyota Camry, Nissan Altima,Nissan Sentra and Chevrolet with 135, 82, 72, 69 respectively.The vehicles were headed  straight ahead,turning  right,slow in traffic,straigh ahead and parked respectively. Hence ensure all road signs are well and clearly  labeled .